### PR TITLE
[MI-2366]:Fixed review fixes of confluence PR 88 on mattermost

### PR DESCRIPTION
**Pull request by Kshitij-Katiyar**

### CODE REVIEW

File: server/command.go [hunk 1]
```diff
@@ -603,6 +603,30 @@ func getFullHelpText(p *Plugin, context *model.CommandArgs, args ...string) stri
 	return helpText
 }
 
+func (p *Plugin) executeConnectInfo(userID, confluenceURL string) (types.ID, string) {
+	info, err := p.GetUserInfo(types.ID(userID), nil)
+	if err != nil {
+		return "", fmt.Sprintf("Failed to connect. Error: %v", err)
+	}
+	if info.Instances.IsEmpty() {
```
Comment (ayusht2810) [inline server/command.go]:
```suggestion
	}

	if info.Instances.IsEmpty() {
```

File: server/service/confluence_service.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+type ConfluenceStatus struct {
+	State string `json:"state"`
+}
+
+func NormalizeConfluenceURL(confluenceURL string) (string, error) {
+	u, err := url.Parse(confluenceURL)
+	if err != nil {
+		return "", err
+	}
+	if u.Host == "" {
```
Comment (ayusht2810) [inline server/service/confluence_service.go]:
```suggestion
	}

	if u.Host == "" {
```

Comment (ayusht2810) [reply server/service/confluence_service.go]: check everywhere

File: server/command.go (hunk 1 above)
Comment (ayusht2810) [reply server/command.go]: check everywhere

File: server/service/confluence_service.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	if u.Scheme == "" {
+		u.Scheme = "https"
+	}
+
+	confluenceURL = strings.TrimSuffix(u.String(), "/")
+	return confluenceURL, nil
+}
+
+// CheckConfluenceURL checks if `/status` endpoint of the Confluence URL is accessible
+// and responding with the correct state which is "RUNNING"
```
Comment (ayusht2810) [inline server/service/confluence_service.go]:
```suggestion
// CheckConfluenceURL checks if the `/status` endpoint of the Confluence URL is accessible
// and responding with the correct state which is "RUNNING"
```

File: server/service/confluence_service.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+func CheckConfluenceURL(mattermostSiteURL, confluenceURL string, requireHTTPS bool) (_ string, err error) {
+	confluenceURL, err = NormalizeConfluenceURL(confluenceURL)
+	if err != nil {
+		return "", err
+	}
+	if confluenceURL == strings.TrimSuffix(mattermostSiteURL, "/") {
```
Comment (ayusht2810) [inline server/service/confluence_service.go]: same

File: server/service/confluence_service.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	contentType := "application/json"
+	buf := &bytes.Buffer{}
+	err = json.NewEncoder(buf).Encode(in)
+	if err != nil {
+		return nil, err
+	}
+	return call(url, method, path, contentType, buf, out, httpClient)
+}
+
+func call(basePath, method, path, contentType string, inBody io.Reader, out interface{}, httpClient *http.Client) (responseData []byte, err error) {
```
Comment (ayusht2810) [inline server/service/confluence_service.go:115]: don't we need to return the status code from this function?

File: server/service/confluence_service.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	case http.StatusNoContent:
+		return nil, nil
+
+	case http.StatusNotFound:
+		return nil, errors.Errorf(ErrorStatusNotFound)
+	}
+
+	type ErrorResponse struct {
+		Message string `json:"message"`
+	}
```
Comment (ayusht2810) [inline server/service/confluence_service.go]: why it is declared here?

File: server/service/confluence_service.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+
+	resBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	var status ConfluenceStatus
+	err = json.Unmarshal(resBody, &status)
+	if err != nil {
+		return "", err
+	}
```
Comment (ayusht2810) [inline server/service/confluence_service.go]:
```suggestion
	if err = json.Unmarshal(resBody, &status); err != nil {
		return "", err
	}
```

File: server/service/confluence_service.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+	return endpointURL.String(), nil
+}
+
+func CallJSON(url, method, path string, in, out interface{}, httpClient *http.Client) (responseData []byte, err error) {
+	contentType := "application/json"
+	buf := &bytes.Buffer{}
+	err = json.NewEncoder(buf).Encode(in)
+	if err != nil {
+		return nil, err
+	}
```
Comment (ayusht2810) [inline server/service/confluence_service.go]:
```suggestion
	if err = json.NewEncoder(buf).Encode(in); err != nil {
		return nil, err
	}
```

File: server/service/confluence_service.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+	type ErrorResponse struct {
+		Message string `json:"message"`
+	}
+	errResp := ErrorResponse{}
+	err = json.Unmarshal(responseData, &errResp)
+	if err != nil {
+		return nil, err
+	}
```
Comment (ayusht2810) [inline server/service/confluence_service.go]: same

**Review by ayusht2810 — APPROVED** [review]

File: server/service/confluence_service.go [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+	resBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	var status ConfluenceStatus
+	if err = json.Unmarshal(resBody, &status); err != nil {
+		return "", err
+	}
```
Comment (avas27JTG) [inline server/service/confluence_service.go]: should we do all this using `callJSON` and return the unmarshalled struct here?

File: server/service/confluence_service_test.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGetEndpointURL(t *testing.T) {
+	for name, val := range map[string]struct {
+		instanceURL    string
+		path           string
+		expectedResult string
+	}{
+		"get enpoint url": {
```
Comment (avas27JTG) [inline server/service/confluence_service_test.go]:
```suggestion
		"get endpoint URL": {
```

File: server/command.go [hunk 12]
```diff
@@ -603,6 +603,34 @@ func getFullHelpText(p *Plugin, context *model.CommandArgs, args ...string) stri
...
+
+	if confluenceURL == "" {
+		if info.connectable.Len() == 1 {
+			confluenceURL = info.connectable.IDs()[0].String()
+		}
+	}
+
+	instanceID := types.ID(confluenceURL)
+	if info.connectable.IsEmpty() {
+		return instanceID, fmt.Sprintf("You already have connected all available Confluence accounts. Please use `/confluence disconnect --instance=%s` to disconnect.", instanceID)
```
Comment (raghavaggarwal2308) [inline server/command.go]:
```suggestion
		return instanceID, fmt.Sprintf("You have already connected all available Confluence accounts. Please use `/confluence disconnect --instance=%s` to disconnect.", instanceID)
```

**Review by raghavaggarwal2308 — APPROVED** [review]

**Pull request merged**